File: benchmarks/solutions/go/r2-json-transform.go
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"os"
)

type User struct {
	Name   string `json:"name"`
	Age    int    `json:"age"`
	Active bool   `json:"active"`
	Email  string `json:"email"`
}

type Output struct {
	Name  string `json:"name"`
	Email string `json:"email"`
}

func main() {
	data, _ := io.ReadAll(os.Stdin)
	var users []User
	if err := json.Unmarshal(data, &users); err != nil {
		fmt.Fprintf(os.Stderr, "error: %v\n", err)
		os.Exit(1)
	}

	var result []Output
	for _, u := range users {
		if u.Active && u.Age > 18 {
			result = append(result, Output{Name: u.Name, Email: u.Email})
		}
	}

	out, _ := json.Marshal(result)
	fmt.Println(string(out))
}

File: benchmarks/solutions/haskell/r2-json-transform.hs
module Main where

import Data.Aeson
import Data.Aeson.Types (parseMaybe)
import qualified Data.ByteString.Lazy as BL
import qualified Data.Vector as V
import System.Exit (exitFailure)
import System.IO (hPutStrLn, stderr)

data User = User { name :: String, age :: Int, active :: Bool, email :: String }

instance FromJSON User where
  parseJSON = withObject "User" $ \v ->
    User <$> v .: "name" <*> v .: "age" <*> v .: "active" <*> v .: "email"

data Output = Output { outName :: String, outEmail :: String }

instance ToJSON Output where
  toJSON (Output n e) = object ["name" .= n, "email" .= e]

main :: IO ()
main = do
  input <- BL.getContents
  case eitherDecode input :: Either String [User] of
    Left err -> hPutStrLn stderr ("error: " ++ err) >> exitFailure
    Right users -> do
      let result = [ Output (name u) (email u)
                   | u <- users, active u, age u > 18 ]
      BL.putStr (encode result)
      putStrLn ""
